packages/remix-oxygen/src/server.ts

```diff
@@ -3,23 +3,63 @@ import {
   type AppLoadContext,
   type ServerBuild,
 } from '@remix-run/server-runtime';
+import {
+  cacheResponse,
+  getItemFromCache,
+  isPublicCacheControlHeader,
+  isStale,
+  shouldCacheRequest,
+} from './cache';
+
+type CreateOxygenRequestHandler = {
+  build: ServerBuild;
+  mode?: string;
+  getLoadContext?: (
+    request: Request,
+  ) => Promise<AppLoadContext> | AppLoadContext;
+  waitUntil?: ExecutionContext['waitUntil'];
+};
 
-export function createRequestHandler<Context = unknown>({
+export function createRequestHandler({
   build,
   mode,
   getLoadContext,
-}: {
-  build: ServerBuild;
-  mode?: string;
-  getLoadContext?: (request: Request) => Promise<Context> | Context;
-}) {
+  waitUntil,
+}: CreateOxygenRequestHandler) {
   const handleRequest = createRemixRequestHandler(build, mode);
+  const handleRequestWithContext = async (request: Request) =>
+    handleRequest(request, await getLoadContext?.(request));
 
   return async (request: Request) => {
-    return handleRequest(
-      request,
-      (await getLoadContext?.(request)) as AppLoadContext,
-    );
+    const cache =
+      !!waitUntil &&
+      shouldCacheRequest(request) &&
+      (await caches.open('oxygen'));
+
+    if (cache) {
+      const cachedResponse = await getItemFromCache(cache, request);
+
+      if (cachedResponse) {
+        if (isStale(request, cachedResponse.headers)) {
+          cacheResponse(
+            request,
+            () => handleRequestWithContext(request),
+            cache,
+            waitUntil,
+          );
+        }
+
+        return cachedResponse;
```

Comment:
Does it mean we return the cached response nonetheless, even if it's stale (see the conditional above)?